tq/ssh.go

```diff
@@ -0,0 +1,391 @@
+package tq
+
+import (
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"sort"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/git-lfs/git-lfs/errors"
+	"github.com/git-lfs/git-lfs/lfshttp"
+	"github.com/git-lfs/git-lfs/ssh"
+	"github.com/git-lfs/git-lfs/tools"
+	"github.com/rubyist/tracerx"
+)
+
+type SSHBatchTransferAdapter struct {
+	maxRetries int
+	transfer   *ssh.SSHTransfer
+}
+
+func (a *SSHBatchTransferAdapter) batchInternal(args []string, batchLines []string) (int, []string, []string, error) {
+	conn := a.transfer.Connection()
+	conn.Lock()
+	defer conn.Unlock()
+	err := conn.SendMessageWithArguments("batch", args, batchLines)
+	if err != nil {
+		return 0, nil, nil, errors.Wrap(err, "batch request")
+	}
+
+	status, args, lines, err := conn.ReadStatusWithArguments()
+	if err != nil {
+		return status, nil, nil, errors.Wrap(err, "batch response")
+	}
+	return status, args, lines, err
+}
+
+func (a *SSHBatchTransferAdapter) Batch(remote string, bReq *batchRequest) (*BatchResponse, error) {
+	bRes := &BatchResponse{TransferAdapterName: "ssh"}
+	if len(bReq.Objects) == 0 {
+		return bRes, nil
+	}
+
+	missing := make(map[string]bool)
+	batchLines := make([]string, 0, len(bReq.Objects))
+	for _, obj := range bReq.Objects {
+		missing[obj.Oid] = obj.Missing
+		batchLines = append(batchLines, fmt.Sprintf("%s %d", obj.Oid, obj.Size))
+	}
+
+	tracerx.Printf("api: batch %d files", len(bReq.Objects))
+
+	requestedAt := time.Now()
+	args := []string{"transfer=ssh"}
+	if bReq.Ref != nil {
+		args = append(args, fmt.Sprintf("refname=%s", bReq.Ref.Name))
+	}
+	status, args, lines, err := a.batchInternal(args, batchLines)
+	if err != nil {
+		return nil, err
+	}
+
+	if status != 200 {
+		msg := "no message provided"
+		if len(lines) > 0 {
+			msg = lines[0]
+		}
+		return nil, fmt.Errorf("batch response: status %d from server (%s)", status, msg)
+	}
+
+	sort.Strings(lines)
+	for _, line := range lines {
+		entries := strings.Split(line, " ")
+		if len(entries) < 3 {
+			return nil, fmt.Errorf("batch response: malformed response")
+		}
+		length := len(bRes.Objects)
+		if length == 0 || bRes.Objects[length-1].Oid != entries[0] {
+			bRes.Objects = append(bRes.Objects, &Transfer{Actions: make(map[string]*Action)})
+		}
+		transfer := bRes.Objects[len(bRes.Objects)-1]
+		transfer.Oid = entries[0]
+		transfer.Size, err = strconv.ParseInt(entries[1], 10, 64)
+		if err != nil {
+			return nil, fmt.Errorf("batch response: invalid size: %s", entries[1])
+		}
```

Comment:
I don't think we can do that because your proposed line 82 essentially calls `bRes.Objects[-1]` if `length` is 0, and it will also overwrite the previous item if this line starts a new stanza.  At least, I believe it does so, although if I'm misunderstanding, do say so.

Reply:
No, that makes sense, my mistake!